pkg/intermediate/aggregate.go

```diff
@@ -124,74 +140,272 @@ func (a *AggregationProcess) DeleteFlowKeyFromMap(flowKey FlowKey) {
-					isFilled = true
-				}
+
+	aggregationRecord, exist := a.flowKeyRecordMap[*flowKey]
+	if exist {
+		if !isRecordIntraNode(record) {
+			// Do correlation of records if record belongs to inter-node flow and
+			// records from source and destination node are not received.
+			if !aggregationRecord.ReadyToSend && !areRecordsFromSameNode(record, aggregationRecord.Record) {
+				a.correlateRecords(record, aggregationRecord.Record)
+				aggregationRecord.ReadyToSend = true
 			}
-			if isFilled {
-				break
+			// Aggregation of incoming flow record with existing by updating stats
+			// and flow timestamps.
+			if isRecordFromSrc(record) {
+				a.aggregateRecords(record, aggregationRecord.Record, true, false)
+			} else {
+				a.aggregateRecords(record, aggregationRecord.Record, false, true)
 			}
+		} else {
+			// For intra-node flows, just do aggregation of the flow record with
+			// existing record by updating the stats and flow timestamps. Correlation
+			// is not required.
+			a.aggregateRecords(record, aggregationRecord.Record, true, true)
 		}
 	} else {
-		for _, existingRec := range existingRecords {
-			if isRecordFromSrc(existingRec) {
-				for _, field := range a.correlateFields {
-					if ieWithValue, exist := record.GetInfoElementWithValue(field); exist {
-						existingIeWithValue, _ := existingRec.GetInfoElementWithValue(field)
-						existingIeWithValue.Value = ieWithValue.Value
+		// Add all the new stat fields and initialize them.
+		if !isRecordIntraNode(record) {
+			if isRecordFromSrc(record) {
+				if err := a.addFieldsForStatsAggregation(record, true, false); err != nil {
+					return err
+				}
+			} else {
+				if err := a.addFieldsForStatsAggregation(record, false, true); err != nil {
+					return err
+				}
+			}
+		} else {
+			if err := a.addFieldsForStatsAggregation(record, true, true); err != nil {
+				return err
+			}
+		}
+		aggregationRecord = AggregationFlowRecord{
+			record,
+			false,
+			true,
+		}
+		if isRecordIntraNode(record) {
+			aggregationRecord.ReadyToSend = true
+		}
+	}
+
+	a.flowKeyRecordMap[*flowKey] = aggregationRecord
+	return nil
+}
+
+// correlateRecords correlate the incomingRecord with existingRecord using correlation
+// fields.
+func (a *AggregationProcess) correlateRecords(incomingRecord, existingRecord entities.Record) {
+	for _, field := range a.correlateFields {
+		if ieWithValue, exist := incomingRecord.GetInfoElementWithValue(field); exist {
+			switch ieWithValue.Element.DataType {
+			case entities.String:
+				if ieWithValue.Value != "" {
+					existingIeWithValue, _ := existingRecord.GetInfoElementWithValue(field)
+					if existingIeWithValue.Value != "" {
+						klog.Warningf("This field with name %v should not have been filled with value %v in existing record.", field, existingIeWithValue.Value)
 					}
+					existingIeWithValue.Value = ieWithValue.Value
 				}
+			case entities.Unsigned16:
+				if ieWithValue.Value != uint16(0) {
```

Comment:
Could 0 be a valid value??

Reply:
Yes. For example, for pod-to-pod flow the destinationServicePort (k8s service port) is not set.

Reply:
If 0 could be a valid value in some cases, do we still skip the storing when the incoming record has 0 value ? Or the 0 value actually has no useful info?

Reply:
No we do not skip storing it, we store everything sent by exporter at collector. If there is a field in IPFIX template, the exporter has to send some value, otherwise collector cannot validate the corresponding IPFIX data record. 0 value of course do not have useful info, but also needed in some case where it has to be correlated with its corresponding record and filled up. 

We have to use and manage multiple templates if the field has no use for specific use case for example destinationServicePortName/destinationService are not applicable for Pod-To-Pod flows. Currently, we use only one template in Antrea, the user of go-ipfix library.

Reply:
I guess here the default value for unint16 type is 0, so only a non-zero value from incoming record needs to be stored.

Reply:
Yes, we only need to correlate non-zero uint16 value and fill up the existing record in flowKeyRecord map.